### PR TITLE
Impl `From<core::convert::Infallible>` for Errors

### PR DESCRIPTION
**Pull request by liamaharon**
Closes https://github.com/rust-bitcoin/rust-bitcoin/issues/1222

Example usage
```rust
use internals::impl_from_infallible;

#[derive(Debug, Clone, PartialEq, Eq)]
enum Error {
    SomeError(u8),
    SomeOtherError(u32, u32),
}

impl_from_infallible!(Error);
```

Feel free to request changes :)

### CODE REVIEW

**Comment by apoelstra** [comment]
I'm torn. This looks well-written at first glance and we've talked for a long time about having proc-macros. On the other hand `quote` and `syn` have caused us a lot of grief, and they also increase the compile time a fair bit. On the third hand, while I don't like proc-macros due to their unreadability, this one is extraordinarily readable (though it would be less so without `quote`). On the fourth hand, it could be done even-more-readably using a normal `macro_rules` macro.

Would you be willing to retry this PR without using syn/quote and see how bad it is?

CI is not going to pass unless you update the included lockfiles to add all the new dependencies.

**Comment by apoelstra** [comment]
Can you rebase rather than adding a commit? I don't mind re-reviewing the whole PR but it's hard to review if the final commit undoes stuff from the first one.

**Comment by liamaharon** [comment]
I'm not so familiar with rebasing but think I did it, let me know if that's better.

btw, which rust version do you use for formatting? I noticed some of the `rustfmt.toml` options are only on nightly, but running `cargo +nightly fmt` made a lot of changes.

**Comment by apoelstra** [comment]
@liamaharon we use nightly for formatting but we do it weekly via cronjob rather than requiring developers run nightly on their own machines (which aside from general concerns with Mozilla and Amazon, requires an ongoing trust relationship with rustup, which for the longest time did not even sign the binaries it was downloading).

But the weekly job hasn't been run for several weeks because @Kixunil did not want to keep rebasing his own stuff.

**Comment by tcharding** [comment]
To get past CI you'll need the following steps (or you can just grab the patch from my fork on branch `rebase-strip-formatting-2516`, its still your patch).

- Stash the current commit someplace (ether with a tmp branch of `git format-patch`)
- Run the formatter and commit the changes `cargo +nightly fmt`
- Apply the patch you stashed (there is one conflict)
- Update the lockfiles by running `just update-lock-files`, and squash the changes into the top commit you just applied
- Rebase and remove the formatting patch (using `git rebase -i master`), there is one conflict same place as above.
- PROFIT

FTR this is an embarrassingly complicated series of steps required to accept contributions from outside developers ...

**Comment by tcharding** [comment]
The PR title is stale also mate.

**Comment by tcharding** [comment]
I guess that is something to do with your environment, I'd prefer to leave it as `--force`.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -309,6 +309,7 @@ More specifically an error should
 - derive `Debug, Clone, PartialEq, Eq` (and `Copy` iff not `non_exhaustive`).
 - implement Display using `write_err!()` macro if a variant contains an inner error source.
 - have `Error` suffix
+- implement `internals::impl_from_infallible`.
```
Comment (tcharding) [inline CONTRIBUTING.md:312]:
```suggestion
- call `internals::impl_from_infallible!
```

File: base58/src/lib.rs [hunk 2]
```diff
@@ -24,13 +24,14 @@ extern crate std;
...
 
-use core::{fmt, iter, slice, str};
 #[cfg(not(feature = "std"))]
 pub use alloc::{string::String, vec::Vec};
+use core::{fmt, iter, slice, str};
 #[cfg(feature = "std")]
 pub use std::{string::String, vec::Vec};
 
 use hashes::{sha256d, Hash};
+use internals::impl_from_infallible;
```
Comment (tcharding) [inline base58/src/lib.rs:34]: You could just use the fully qualified path if you want to then you don't have to touch import statements at all, may make your life easier.

File: bitcoin/src/address/mod.rs [hunk 3]
```diff
@@ -49,8 +50,6 @@ use crate::crypto::key::{
 use crate::network::{Network, NetworkKind};
 use crate::prelude::*;
 use crate::taproot::TapNodeHash;
-
-use self::error::P2shError;
```
Comment (tcharding) [inline bitcoin/src/address/mod.rs:53]: This PR still has formatting changes in it, can we remove them please?

Comment (liamaharon) [reply bitcoin/src/address/mod.rs:53]: Oh oops, I missed your instructions to remove the formatting patch. Let me try again.

Comment (tcharding) [reply bitcoin/src/address/mod.rs:53]: Yeah sorry man, this whole formatter thing is an unsolved problem - we are trying our best to keep everyone happy, its no easy task.

**Comment by tcharding** [comment]
Also, massive props for reading the contributor docs and following the guidelines!

**Comment by tcharding** [comment]
I'm not sure why this PR is triggering the CI fail but I fixed it in #2521.

File: bitcoin/src/address/mod.rs (hunk 3 above)
Comment (liamaharon) [reply bitcoin/src/address/mod.rs]: No worries ;) tricky problem when not using rustup.

Force pushed again, diff much more manageable now.

**Comment by tcharding** [comment]
Oh, I just noticed, this does not do all errors in the codebase. Perhaps either change the commit log to say "implement `Infallible` for all errors called `Error`" or update to do all the others.

**Comment by liamaharon** [comment]
> Oh, I just noticed, this does not do all errors in the codebase. Perhaps either change the commit log to say "implement `Infallible` for all errors called `Error`" and don't close the issue, or update to do all the errors.

Good catch @tcharding, my mistake for missing those. 

I've updated the macro to support types with generics and lifetimes, and walked through all enum definitions applying the macro where appropriate.

The only type I'm not sure how to handle is `FromHexError<E>`, because it already has a blanket `From` implementation for all `E`: `impl<E> From<E> for FromHexError<E>`. I've left a note in the doccomment for now, but let me know if you'd like to handle it some other way.

File: internals/src/macros.rs [hunk 4]
```diff
@@ -113,3 +113,57 @@ macro_rules! const_assert {
...
+///     what: &'b D,
+/// }
+/// impl_from_infallible!(DeltaStruct<'b, 'a: 'static + 'b, T: 'a, D: Debug + Display + 'a>);
+/// ```
+///
+/// See <https://stackoverflow.com/a/61189128> for more information about this macro.
+#[macro_export]
+macro_rules! impl_from_infallible {
+    ( $name:ident $(< $( $lt:tt $( : $clt:tt $(+ $dlt:tt )* )? ),+ >)? ) => {
+        impl $(< $( $lt $( : $clt $(+ $dlt )* )? ),+ >)?
```
Comment (tcharding) [inline internals/src/macros.rs:161]: Totally unreadable but awesome docs/doctests - works for me!

**Review by tcharding — DISMISSED** [review]
Thanks for sticking with it man, like a boss!

ACK d9b3dc6c1e1073163650838d9ccfdedabad4526b

File: internals/src/macros.rs (hunk 4 above)
Comment (liamaharon) [reply internals/src/macros.rs:163]: Yeah :D

A proc macro would be a more readable, but as you say worse for compile performance.

**Comment by liamaharon** [comment]
No worries @tcharding! Made a tiny doc tweak, needs re-review please.

File: internals/src/macros.rs (hunk 4 above)
Comment (Kixunil) [reply internals/src/macros.rs:163]: > A proc macro would be a more readable

I don't think so. I find proc maros less readable because they say every detail of how they do things instead of being declarative. So my head has to interpret the code and infer the intent from what it's doing.

File: bitcoin/src/string.rs [hunk 5]
```diff
@@ -35,6 +35,9 @@ pub trait FromHexStr: Sized {
 }
 
 /// Hex parsing error
+///
+/// Does not implement [`internals::impl_from_infallible`] because there is already a blanket `impl
+/// From<E> for FromHexError<E>` defined which converts the error into `FromHexError::ParseHex(E)`.
```
Comment (Kixunil) [inline bitcoin/src/string.rs:40]: I think it's fine. I believe conversion from infallible is more rare than conversion from inner error type.

**Review by Kixunil — DISMISSED** [review]
ACK 715bb5ee36ea5934289f123045025d43964fd56b

**Review by tcharding — APPROVED** [review]
ACK 715bb5ee36ea5934289f123045025d43964fd56b

**Comment by apoelstra** [comment]
CI failures look real, and also this needs a rebase.

**Comment by liamaharon** [comment]
@apoelstra are the CI failures not unrelated and fixed in https://github.com/rust-bitcoin/rust-bitcoin/pull/2521?

**Comment by apoelstra** [comment]
Oh, you're correct, it looks like the failing checks now are all related to #2521.

**Review by apoelstra — APPROVED** [review]
ACK b9f7462958ca1d47692938fc8c6292569823077f

**Review by Kixunil — APPROVED** [review]
ACK b9f7462958ca1d47692938fc8c6292569823077f
